File: target-web/src/semantik.js
'use strict';
// ===========================================================================
// TAHAP 4 — PENGANALISIS SEMANTIK
// Memeriksa makna: wadah dipakai sebelum disiapkan, kebiasaan ada/tidak,
// jumlah isian cocok, 'mengembalikan' hanya di dalam kebiasaan, dsb.
// ===========================================================================

class Penganalisis {
  constructor() {
    this.diagnostik = [];
    this.kerangka = [new Map()]; // tumpukan lingkup: Map(nama -> {simpul, dipakai})
    this.fungsi = new Map();     // nama -> {jumlah, simpul, dipakai}
    this.dalamFungsi = 0;
    this.jumlahBab = 0;
    this.jumlahKalimat = 0;
  }

  galat(simpul, pesan) {
    this.diagnostik.push({ tingkat: 'galat', pesan, baris: simpul.baris, berkas: simpul.berkas });
  }
  peringatan(simpul, pesan) {
    this.diagnostik.push({ tingkat: 'peringatan', pesan, baris: simpul.baris, berkas: simpul.berkas });
  }

  // -- pengelolaan lingkup --------------------------------------------------
  bawah() { return this.kerangka[this.kerangka.length - 1]; }
  dorong() { this.kerangka.push(new Map()); }
  tarik(simpul) {
    const f = this.kerangka.pop();
    for (const [nama, info] of f) {
      if (!info.dipakai) {
        this.peringatan(info.simpul, `Wadah '${nama}' disiapkan tetapi tidak pernah dipakai.`);
      }
    }
    if (simpul) { /* placeholder */ }
  }
  deklare(nama, simpul) { this.bawah().set(nama, { simpul, dipakai: false }); }
  ada(nama) {
    for (let i = this.kerangka.length - 1; i >= 0; i--) if (this.kerangka[i].has(nama)) return true;
    return false;
  }
  tandai(nama) {
    for (let i = this.kerangka.length - 1; i >= 0; i--) {
      if (this.kerangka[i].has(nama)) { this.kerangka[i].get(nama).dipakai = true; return; }
    }
  }

  // -- pemeriksaan ungkapan -------------------------------------------------
  periksaUngkapan(n, simpul) {
    if (!n) return;
    switch (n.t) {
      case 'Angka': case 'Teks': case 'Benar': case 'Salah': return;
      case 'Wadah': case 'Panjang': case 'KeAngka': case 'HurufKecil': case 'HurufBesar':
        if (!this.ada(n.nama)) {
          this.galat(simpul, `Wadah '${n.nama}' dipakai sebelum pernah disiapkan.`);
        } else this.tandai(n.nama);
        return;
      case 'Panggilan':
        for (const a of n.args) this.periksaUngkapan(a, simpul);
        this.periksaPanggilan(n.nama, n.args.length, simpul);
        return;
      case 'Biner': case 'Logika':
        this.periksaUngkapan(n.kiri, simpul); this.periksaUngkapan(n.kanan, simpul);
        return;
      case 'Tidak': this.periksaUngkapan(n.anak, simpul); return;
      case 'Kondisi': case 'Mengandung':
        this.periksaUngkapan(n.kiri, simpul); this.periksaUngkapan(n.kanan, simpul);
        return;
      case 'Kebenaran': this.periksaUngkapan(n.expr, simpul); return;
      default: return;
    }
  }

  periksaPanggilan(nama, jumlahArg, simpul) {
    if (!this.fungsi.has(nama)) {
      this.galat(simpul, `Kebiasaan '${nama}' dipanggil tetapi belum pernah dibuat.`);
      return;
    }
    const f = this.fungsi.get(nama);
    f.dipakai = true;
    if (f.jumlah !== jumlahArg) {
      this.galat(simpul, `Kebiasaan '${nama}' menerima ${f.jumlah} isian, tetapi diberi ${jumlahArg}.`);
    }
  }

  // -- penelusuran AST ------------------------------------------------------
  analisis(program) {
    for (const n of program.isi) this.kalimat(n);
    this.tarik(null);
    for (const [nama, info] of this.fungsi) {
      if (!info.dipakai) this.peringatan(info.simpul, `Kebiasaan '${nama}' dibuat tetapi tidak pernah dipanggil.`);
    }
    return {
      diagnostik: this.diagnostik,
      ringkasan: {
        bab: this.jumlahBab,
        kalimat: this.jumlahKalimat,
        fungsi: this.fungsi.size,
        galat: this.diagnostik.filter((d) => d.tingkat === 'galat').length,
        peringatan: this.diagnostik.filter((d) => d.tingkat === 'peringatan').length,
      },
    };
  }

  kalimat(n) {
    this.jumlahKalimat++;
    switch (n.t) {
      case 'Bab': this.jumlahBab++; return;
      case 'Let':
        this.periksaUngkapan(n.nilai, n); this.deklare(n.nama, n); return;
      case 'Set': case 'Tambah': case 'Kurang':
        if (!this.ada(n.nama)) this.galat(n, `Wadah '${n.nama}' diubah sebelum pernah disiapkan.`);
        else this.tandai(n.nama);
        this.periksaUngkapan(n.nilai, n); return;
      case 'Berkata': this.periksaUngkapan(n.nilai, n); return;
      case 'Tanya': this.periksaUngkapan(n.pertanyaan, n); this.deklare(n.nama, n); return;
      case 'Jika':
        this.periksaUngkapan(n.kondisi, n);
        for (const a of n.tubuh) this.kalimat(a);
        for (const a of n.lain) this.kalimat(a);
        return;
      case 'Selama':
        this.periksaUngkapan(n.kondisi, n);
        for (const a of n.tubuh) this.kalimat(a);
        return;
      case 'Kebiasaan': {
        this.fungsi.set(n.nama, { jumlah: n.params.length, simpul: n, dipakai: false });
        this.dorong();
        for (const p of n.params) this.deklare(p, n);
        this.dalamFungsi++;
        for (const a of n.tubuh) this.kalimat(a);
        this.dalamFungsi--;
        this.tarik(n);
        return;
      }
      case 'Panggil':
        for (const a of n.args) this.periksaUngkapan(a, n);
        this.periksaPanggilan(n.nama, n.args.length, n);
        return;
      case 'Kembali':
        if (this.dalamFungsi === 0) this.galat(n, "'mengembalikan' hanya boleh dipakai di dalam sebuah kebiasaan.");
        this.periksaUngkapan(n.nilai, n);
        return;
      case 'Sambung':
        this.galat(n, `Penyambungan cerita '${n.berkas}' belum diselesaikan.`);
        return;
      case 'Halaman': this.periksaUngkapan(n.judul, n); return;
      case 'Judul': case 'Paragraf': case 'Jawaban': case 'GantiJawaban':
        this.periksaUngkapan(n.nilai, n); return;
      case 'Isian': return;
      case 'Tombol':
        this.periksaUngkapan(n.label, n);
        if (!this.fungsi.has(n.fungsi)) {
          this.galat(n, `Tombol memanggil kebiasaan '${n.fungsi}' yang belum pernah dibuat.`);
        } else this.fungsi.get(n.fungsi).dipakai = true;
        return;
      case 'BacaIsian': this.deklare(n.sasaran, n); return;
      case 'Ambil': this.deklare(n.nama, n); return;
      case 'Serahkan':
        if (!this.fungsi.has(n.nama)) {
          this.galat(n, `Kebiasaan '${n.nama}' diserahkan ke dunia tetapi belum pernah dibuat.`);
        } else this.fungsi.get(n.nama).dipakai = true;
        return;
      default: return;
    }
  }
}

function analisis(program) {
  return new Penganalisis().analisis(program);
}

module.exports = { Penganalisis, analisis };

File: src/insjayvm/semantik.py
class Penganalisis:
    BAWAAN = {
        "akar": 1, "pangkat": 2, "bulat": 1, "lantai": 1, "acak": 0,
        "waktu_sekarang": 0,
    }

    def __init__(self):
        self.diagnostik = []
        self.kerangka = [{}]
        self.fungsi = {}
        self.dalam_fungsi = 0
        self.jumlah_bab = 0
        self.jumlah_kalimat = 0

    def galat(self, n, pesan):
        self.diagnostik.append({"tingkat": "galat", "pesan": pesan,
                                "baris": n.get("baris"), "berkas": n.get("berkas")})

    def peringatan(self, n, pesan):
        self.diagnostik.append({"tingkat": "peringatan", "pesan": pesan,
                                "baris": n.get("baris"), "berkas": n.get("berkas")})

    def bawah(self):
        return self.kerangka[-1]

    def dorong(self):
        self.kerangka.append({})

    def tarik(self):
        f = self.kerangka.pop()
        for nama, info in f.items():
            if not info["dipakai"]:
                self.peringatan(info["simpul"], "Wadah '%s' disiapkan tetapi tidak pernah dipakai." % nama)

    def deklare(self, nama, simpul):
        self.bawah()[nama] = {"simpul": simpul, "dipakai": False}

    def ada(self, nama):
        return any(nama in f for f in self.kerangka)

    def tandai(self, nama):
        for f in reversed(self.kerangka):
            if nama in f:
                f[nama]["dipakai"] = True
                return

    def periksa_ungkapan(self, n, asal):
        if n is None:
            return
        t = n["t"]
        if t in ("Angka", "Teks", "Benar", "Salah"):
            return
        if t in ("Wadah", "Panjang", "KeAngka", "HurufKecil", "HurufBesar"):
            if not self.ada(n["nama"]):
                self.galat(asal, "Wadah '%s' dipakai sebelum pernah disiapkan." % n["nama"])
            else:
                self.tandai(n["nama"])
        elif t == "Panggilan":
            for a in n["args"]:
                self.periksa_ungkapan(a, asal)
            self.periksa_panggilan(n["nama"], len(n["args"]), asal)
        elif t in ("Biner", "Logika"):
            self.periksa_ungkapan(n["kiri"], asal)
            self.periksa_ungkapan(n["kanan"], asal)
        elif t == "Tidak":
            self.periksa_ungkapan(n["anak"], asal)
        elif t in ("Kondisi", "Mengandung"):
            self.periksa_ungkapan(n["kiri"], asal)
            self.periksa_ungkapan(n["kanan"], asal)
        elif t == "Kebenaran":
            self.periksa_ungkapan(n["expr"], asal)

    def periksa_panggilan(self, nama, jumlah_arg, asal):
        if nama in self.BAWAAN:
            if self.BAWAAN[nama] != jumlah_arg:
                self.galat(asal, "Pustaka '%s' menerima %d isian, tetapi diberi %d."
                           % (nama, self.BAWAAN[nama], jumlah_arg))
            return
        if nama not in self.fungsi:
            self.galat(asal, "Kebiasaan '%s' dipanggil tetapi belum pernah dibuat." % nama)
            return
        f = self.fungsi[nama]
        f["dipakai"] = True
        if f["jumlah"] != jumlah_arg:
            self.galat(asal, "Kebiasaan '%s' menerima %d isian, tetapi diberi %d."
                       % (nama, f["jumlah"], jumlah_arg))

    def analisis(self, program):
        for n in program["isi"]:
            self.kalimat(n)
        self.tarik()
        for nama, info in self.fungsi.items():
            if not info["dipakai"]:
                self.peringatan(info["simpul"], "Kebiasaan '%s' dibuat tetapi tidak pernah dipanggil." % nama)
        return {
            "diagnostik": self.diagnostik,
            "ringkasan": {
                "bab": self.jumlah_bab,
                "kalimat": self.jumlah_kalimat,
                "fungsi": len(self.fungsi),
                "galat": sum(1 for d in self.diagnostik if d["tingkat"] == "galat"),
                "peringatan": sum(1 for d in self.diagnostik if d["tingkat"] == "peringatan"),
            },
        }

    def kalimat(self, n):
        self.jumlah_kalimat += 1
        t = n["t"]
        if t == "Bab":
            self.jumlah_bab += 1
        elif t == "Let":
            self.periksa_ungkapan(n["nilai"], n)
            self.deklare(n["nama"], n)
        elif t in ("Set", "Tambah", "Kurang"):
            if not self.ada(n["nama"]):
                self.galat(n, "Wadah '%s' diubah sebelum pernah disiapkan." % n["nama"])
            else:
                self.tandai(n["nama"])
            self.periksa_ungkapan(n["nilai"], n)
        elif t == "Berkata":
            self.periksa_ungkapan(n["nilai"], n)
        elif t == "Tanya":
            self.periksa_ungkapan(n["pertanyaan"], n)
            self.deklare(n["nama"], n)
        elif t == "Jika":
            self.periksa_ungkapan(n["kondisi"], n)
            for a in n["tubuh"]:
                self.kalimat(a)
            for a in n["lain"]:
                self.kalimat(a)
        elif t == "Selama":
            self.periksa_ungkapan(n["kondisi"], n)
            for a in n["tubuh"]:
                self.kalimat(a)
        elif t == "Kebiasaan":
            self.fungsi[n["nama"]] = {"jumlah": len(n["params"]), "simpul": n, "dipakai": False}
            self.dorong()
            for p in n["params"]:
                self.deklare(p, n)
            self.dalam_fungsi += 1
            for a in n["tubuh"]:
                self.kalimat(a)
            self.dalam_fungsi -= 1
            self.tarik()
        elif t == "Panggil":
            for a in n["args"]:
                self.periksa_ungkapan(a, n)
            self.periksa_panggilan(n["nama"], len(n["args"]), n)
        elif t == "Kembali":
            if self.dalam_fungsi == 0:
                self.galat(n, "'mengembalikan' hanya boleh dipakai di dalam sebuah kebiasaan.")
            self.periksa_ungkapan(n["nilai"], n)
        elif t == "Sambung":
            self.galat(n, "Penyambungan cerita '%s' belum diselesaikan." % n["berkas"])
        elif t == "Halaman":
            self.periksa_ungkapan(n["judul"], n)
        elif t in ("Judul", "Paragraf", "Jawaban", "GantiJawaban"):
            self.periksa_ungkapan(n["nilai"], n)
        elif t == "Isian":
            pass
        elif t == "Tombol":
            self.periksa_ungkapan(n["label"], n)
            if n["fungsi"] not in self.fungsi:
                self.galat(n, "Tombol memanggil kebiasaan '%s' yang belum pernah dibuat." % n["fungsi"])
            else:
                self.fungsi[n["fungsi"]]["dipakai"] = True
        elif t == "BacaIsian":
            self.deklare(n["sasaran"], n)
        elif t == "Ambil":
            self.deklare(n["nama"], n)
        elif t == "Serahkan":
            if n["nama"] not in self.fungsi:
                self.galat(n, "Kebiasaan '%s' diserahkan ke dunia tetapi belum pernah dibuat." % n["nama"])
            else:
                self.fungsi[n["nama"]]["dipakai"] = True


def analisis(program):
    return Penganalisis().analisis(program)
